core/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -1083,53 +1084,40 @@ public void onMessage(ITransport transport, ProtocolMessage message) throws Ably
     }
 
     private void onChannelMessage(ProtocolMessage message) {
-        if(message.connectionSerial != null) {
-            connection.serial = message.connectionSerial.longValue();
-            if (connection.key != null)
-                connection.recoveryKey = connection.key + ":" + message.connectionSerial;
-        }
         channels.onMessage(message);
     }
 
     private synchronized void onConnected(ProtocolMessage message) {
-        /* if the returned connection id differs from
-         * the existing connection id, then this means
-         * we need to suspend all existing attachments to
-         * the old connection.
-         * If realtime did not reply with an error, it
-         * signifies that this was a result of an earlier
-         * connection being invalidated due to being stale.
-         *
-         * Suspend all channels attached to the previous id;
-         * this will be reattached in setConnection() */
         ErrorInfo error = message.error;
-        if(connection.id != null && !message.connectionId.equals(connection.id)) {
-            /* we need to suspend the original connection */
-            if(error == null) {
-                error = REASON_SUSPENDED;
-            }
-            channels.suspendAll(error, false);
-        }
 
-        /* set the new connection id */
-        ConnectionDetails connectionDetails = message.connectionDetails;
-        connection.key = connectionDetails.connectionKey;
-        if (!message.connectionId.equals(connection.id)) {
-            /* The connection id has changed. Reset the message serial and the
-             * pending message queue (which fails the messages currently in
-             * there). */
-            pendingMessages.reset(msgSerial,
-                    new ErrorInfo("Connection resume failed", 500, 50000));
+        if (message.action == ProtocolMessage.Action.connected && message.connectionId.equals(connection.id) && error == null) {
+            //RTN15c6
+            Log.d(TAG, "connection has reconnected and resumed successfully");
+            connection.reason = null;
+            channels.reAttach();
```

Comment:
My understanding is we will always need to reattach after we get a new connection - what do you think about moving the `channels.reAttach()` after the `if/else` to make that explicit?

(I think it what you've will still work given the spec - though just to be safe incase we get an unexpected response from the backend maybe worth making sure `reAttach` will always run?)

Reply:
As I see now this function does not need `message.action == ProtocolMessage.Action.connected` as this is already filtered to connected events. 
I will refactor this to not check for connected state but I think we need to keep reAttach() in if/else cases as they are specific about connectionId and error received.

Reply:
Though every time we get a new connection we must `reAttach` right? Regardless of whether we get a new connection ID and/or an error? Maybe it doesn't matter - though to me the safest option seems to be move out of the `if/else` so we can be sure we already reattach on a new connection?

Reply:
I am not sure if we should reattach every connect case as spec RTN15c6 and RTN15c7 state when we should do it.

`RTN15c6 - A @CONNECTED@ @ProtocolMessage@ with the same @connectionId@ as the current client (and no @error@ property).`
`RTN15c7 - CONNECTED@ @ProtocolMessage@ with a new @connectionId@ and an @ErrorInfo@ in the @error@ field.`

Reply:
I agree with Andy here, IMO it reads weirdly here that you have the same channels.reattach() line in both branches of a conditional that's intended to be exhaustive

Reply:
Yours suggestion is applied here https://github.com/ably/ably-java/pull/842/commits/a6eeb001531505e02bd5149c2c7864d951d3c203